### PR TITLE
[Intl] Provide translated timezone names

### PR DESCRIPTION
**Pull request by c960657**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR | symfony/symfony-docs#6209 |

This PR adds support for translating timezone names. This is useful e.g. in the TimezoneType form type.

The patch only adds support for identifier (Europe/London) to location/city name mapping (London) and vice versa. I assume this is the most common use-case.

The timezone database contains other elements, including meta-zones, alternative names etc. Support for these may be added in the future.

### CODE REVIEW

File: src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Gives access to locale-related ICU data.
+ */
+interface TimezoneBundleInterface extends ResourceBundleInterface
+{
+    /**
+     * Returns the location name (usually a city) for a timezone.
+     *
+     * @param string $locale        The locale to return the name of (e.g. "de_AT").
```
Comment (Tobion) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: does not correspond to signature

File: src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     *                              Defaults to {@link \Locale::getDefault()}.
+     *
+     * @return string|null The location name for the timezone or NULL if not found.
+     */
+    public function getTimezoneName($timezone, $displayLocale = null);
+
+    /**
+     * Returns the location names (usually cities) for all known timezones.
+     *
+     * @param string $displayLocale Optional. The locale to return the names in.
```
Comment (Tobion) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: string|null

File: src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php (hunk 1 above)
Comment (c960657) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: Fixed.

File: src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php (hunk 2 above)
Comment (c960657) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: @Tobion I'm not sure understand what you are referring to here. AFAICT optional parameters do not include `|null` in the majority Symfony docblocks. The [coding standards](http://symfony.com/doc/current/contributing/code/standards.html) does not mention it explicitly. But perhaps `|null` is the preferred variant nowadays?

Comment (stof) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: We include ``|null`` when passing null is supported. If we only write ``string``, it means that any other value than a string is an invalid usage of the method and so not subject to backward compatibility.

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Converts a timezone identifier to an English string.
+     *
+     * @param string $timezone
+     *
+     * @return string
+     */
+    protected function getFallbackName($timezone)
```
Comment (stof) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: should be private

File: src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php (hunk 2 above)
Comment (c960657) [inline src/Symfony/Component/Intl/ResourceBundle/TimezoneBundleInterface.php]: Ok, thanks for the explanation. I have updated the docblock.

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php (hunk 3 above)
Comment (c960657) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: Fixed.

**Comment by webmozart** [comment]
Thanks for creating this great PR @c960657! :fireworks: I think both the PR and the rest of the Intl component should be improved, API-wise (see #18368). Do you want to work on that?

File: src/Symfony/Component/Intl/Data/Provider/TimezoneDataProvider.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    public function getIDs()
+    {
+        return $this->reader->readEntry($this->path, 'meta', array('Timezones'));
+    }
+
+    public function getName($zoneID, $displayLocale = null)
+    {
+        if (null === $displayLocale) {
+            $displayLocale = Locale::getDefault();
```
Comment (trousers) [inline src/Symfony/Component/Intl/Data/Provider/TimezoneDataProvider.php]: Codacy Issue found: [Avoid using static access to class '\Symfony\Component\Intl\Locale' in method 'getName'.](https://www.codacy.com/app/git_9/symfony/file/2332020092/issues/source?bid=3207821&fileBranchId=3207896#l55)

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+     * @return string
+     */
+    private function getFallbackName($timezone)
+    {
+        $parts = explode('/', $timezone);
+        if (count($parts) > 2) {
+            $name = $parts[2].', '.$parts[1];
+        } elseif (count($parts) > 1) {
+            $name = $parts[1];
+        } else {
```
Comment (trousers) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: Codacy Issue found: [The method getFallbackName uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/git_9/symfony/file/2332020086/issues/source?bid=3207821&fileBranchId=3207896#l157)

File: src/Symfony/Component/Intl/Data/Provider/TimezoneDataProvider.php (hunk 4 above)
Comment (trousers) [inline src/Symfony/Component/Intl/Data/Provider/TimezoneDataProvider.php]: Codacy Issue found: [Avoid using static access to class '\Symfony\Component\Intl\Locale' in method 'getName'.](https://www.codacy.com/app/git_9/symfony/file/2332020092/issues/source?bid=3207821&fileBranchId=3207896#l55)

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php (hunk 5 above)
Comment (trousers) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: Codacy Issue found: [The method getFallbackName uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/git_9/symfony/file/2332020086/issues/source?bid=3207821&fileBranchId=3207896#l157)

Comment (HeahDude) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: @trousers seriously ?

Comment (HeahDude) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: There is almost 10 000 lines here!

**Comment by c960657** [comment]
@webmozart I agree that the Intl component could be improved or perhaps superseded by one or more new components.

It's a bit unclear to me what the purpose of that package is – it provides a copy of a subset of the CLDR datasets (with some opinionated blacklists of pseudo-currencies, pseudo-countries etc.), a partial reimplementation of PHP's ICU DateFormatter for English only, as well as code for parsing ICU data files and converting them to various Symfony-specific file formats.

CLDR contains a **lot** of useful data that I would like to see exposed. Symfony will probably never get support for all the different data, but it should make it easy to fetch the relevant bits from the CLDR data files.

I would be interested in working on this, provided that we can agree on a clear direction. I'll add some comments and suggestions to #18368.

**Comment by stof** [comment]
Is the data contained in this PR build based on the same ICU version than the other data already in the component ?

If yes, this gets a 👍 from me. If no, we need to recompile the data first.

**Comment by curry684** [comment]
Same goes here what I commented in the [other PR](https://github.com/symfony/symfony/pull/17628#issuecomment-290724976): the translation quality appears to be sub par and inconsistent, and in that case, no matter the usefulness of the PR itself, we should not want to do it imho.

**Comment by nicolas-grekas** [comment]
@jakzal could you have a look at this one please? you're the one doing most of the updates :)

**Comment by c960657** [comment]
I don't think it is fair to say that the CLDR translations are poor. CLDR is backed by Microsoft, Google and Apple (among others) who use the data in their software. They all spend considerable ressources on providing high-quality data for a large number of locales.

The blame is on me for using the data the wrong way. Since I submitted this PR last year, I have learned a lot more about [the CLDR data](http://www.unicode.org/repos/cldr/trunk/common/) and [the CLDR spec](http://unicode.org/reports/tr35/) while contributing to the [Punic library](https://github.com/punic/punic/).

The Olson database contains timezone definitions but does not specify how they should be displayed to end users. Symfony simply shows the raw timezone identifier with `/` and `_` replaced with dash and space, respectively. This strategy works most of time for English, but it does not handle apostrophes, periods and accents, e.g. `Antarctica/DumontDUrville` is displayed as “DumontDUrville” rather than “Dumont d’Urville” (search for `examplarCity` in [root.xml](http://www.unicode.org/repos/cldr/trunk/common/main/root.xml) for more examples). So even for English, CLDR may help improve matters.

The Olson database does not designate a name for each timezone apart from the timezone identifier (e.g. `Europe/Berlin`). There are several naming schemes for timezones; for instance, “Berlin time”, “German time” and (to some extent) “CET” are different names for the same timezone. The CLDR spec [defines several naming schemes](http://unicode.org/reports/tr35/tr35-dates.html#Using_Time_Zone_Names), such as _generic non-location format_ and _generic partial location format_, that describes how to combine the CLDR strings into end-user facing strings.

A sidenote, not really related to this issue: If you are creating UI that allows users to select a timezone, showing all 400 Olson timezones is probably not the most user-friendly approach. Many of the Olson timezones are duplicates, in the sense that they (in recent years) have identical offsets and DST rules, e.g. `Europe/Berlin` and `Europe/Paris`. It makes sense to group the zones in the UI, e.g. like in the [Microsoft Windows timezone selector](https://www.microsoft.com/library/media/1033/windowsxp/images/using/setup/tips/67446-change-time-zone.gif). CLDR includes two such groupings; the one used in Windows, and another called _metazones_.


So, how does all this apply to Symfony?

In order to build a multi-locale appliation that handles dates and timezones, you would usually use a l10n library such as [PHPs Intl extension](http://php.net/intl), the Punic library or similar (Symfonys Intl component is not a l10n library in this respect). These packages often provide translated timezone names. E.g. the Intl extension allows you to get the localized city name like this `IntlDateFormatter::formatObject(new DateTime('now', new DateTimeZone('Europe/Rome')), 'VVV', 'it') // returns "Roma"`.

So it seems there is little benefit in adding the timezone names to the Intl package.

(I guess the same applies to the other data currently provided by Symfonys Intl component, but
[as mentioned above](https://github.com/symfony/symfony/pull/17636#issuecomment-204768157), it is unclear to me what the purpose of the Intl package is)


The reason I created this PR in the first place was to allow the timezone form component to be localized (PR #17628). Perhaps a better approach is to use PHPs Intl extension directly at run-time when generating the dropdown (with fallback to English if the Intl extension is not installed). What do you think of this?

**Comment by curry684** [comment]
@c960657 my apologies if I inferred the quality of your work as a whole was poor, I merely reviewed the proposed changes and concluded that the work being submitted here was not generically usable in its current form. And therefore should not (yet) be a core part of Symfony as it would cause technical debt and BC issues if it were corrected later.

**Comment by c960657** [comment]
@curry684 No offense taken :-) It seems we agree that this data is available through other libraries and thus does not belong in Symfony.

**Comment by nicolas-grekas** [comment]
shall we close then?

**Pull request closed without merging**